linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala

```diff
@@ -550,6 +550,12 @@ private final class IRChecker(unit: LinkingUnit, reporter: ErrorReporter) {
       case IdentityHashCode(expr) =>
         typecheckExpr(expr, env)
 
+      case WrapAsThrowable(expr) =>
+        typecheckExpr(expr, env)
+
+      case UnwrapFromThrowable(expr) =>
+        typecheckExpr(expr, env)
```

Comment:
No it doesn't. There's no reason to restrict the type, so why do it? Allowing any value may ease some optimizations in the future, as we could simplify `Unwrap(Wrap(x))` to `Unwrap(x)`. We couldn't do that if it required a `Throwable`.

Reply:
> There's no reason to restrict the type, so why do it?

Hum, could you qualify this argument somewhat? Without additional reasoning, we could apply this to anything, and just type anything as any :-/

> `Unwrap(Wrap(x))` to `Unwrap(x)`

Shouldn't `Unwrap(Wrap(x))` simplify to `x`? Feels like I'm missing something.

In any case, what is the specification of `Unwrap(x)` if `x` is not a `Throwable`? Return `x` I assume?

Reply:
> > There's no reason to restrict the type, so why do it?
> 
> Hum, could you qualify this argument somewhat? Without additional reasoning, we could apply this to anything, and just type anything as any :-/

There are two reasons to restrict the input type of an operation:

1. To ensure soundness of the type system, or
2. To make the implementation of the operation simpler.

Neither of these two reasons apply here:

1. Clearly, type soundness is ensured, since we can write a user-level function that takes an `Any` and does the same thing.
2. AFAICT, it wouldn't be any easier to implement this operation if it had to work only on `Throwable`.

> > `Unwrap(Wrap(x))` to `Unwrap(x)`
> 
> Shouldn't `Unwrap(Wrap(x))` simplify to `x`? Feels like I'm missing something.

Unfortunately not. If `x` is already a `JavaScriptException`, `Wrap(x)` is a no-op, and therefore `Unwrap(x)` returns what's inside the original `x`.

> In any case, what is the specification of `Unwrap(x)` if `x` is not a `Throwable`? Return `x` I assume?

It returns `x`, yes. As tested in `SpecialTest.unwrapFromThrowable`.

Reply:
Is "avoiding mistakes" also a reason to restrict the input type of an operation?

I think both approaches (restrict to the max v.s. restrict as necessary) have their merits. For me this goes a lot into the discussion about what we expect from optimized IR.

IMO an interesting comparison is the `Clone` node, which currently requires its input to be a `jl.Cloneable`. There is no intrinsic reason from an implementation POV that this is necessary. Further, it seems that the optimizer actually breaks this, at least judging from these lines:

https://github.com/scala-js/scala-js/blob/d5b1fa8fa86ae4e91b0fb3b1aa2b19b159d04056/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala#L2620-L2622

I'm unsure about what the best course of action is here. On one hand, it seems like a good idea to be as restrictive as possible for the public IR specification. On the other hand, this means we either generate more invalid IR or need to introduce a mechanism to keep the IR valid (transients with less restrictive typing come to mind). The "lazy" approach we could take is to "ignore" the problem, restrict the input for now and widen later (which we can in a backwards compatible manner IIUC).

Reply:
Right. It's true that there is value in the more restrictive variant.

The `Clone` example is not quite the same because:

* We restricted because we imagined that it would be easier to implement in Wasm, if we ever have a Wasm back-end, as we would avoid cloning metadata for non-`Cloneable` classes.
* The optimizer produces code that's invalid for `Clone` because we don't have a way to carry extra type information from the optimzer to the emitter. But in this case the optimizer *knows* that is a `Cloneable`. At run-time it's always a `Cloneable`. In the hypothetical `Unwrap(Wrap(x))` optimization, the optimizer *wouldn't* know anything about `x`, and `x` could really be anything at run-time.

> The "lazy" approach we could take is to "ignore" the problem, restrict the input for now and widen later (which we can in a backwards compatible manner IIUC).

I was also going to suggest the lazy approach of restricting now and perhaps widen later. So I guess I'll go for that.